pkg/kubelet/token/token_manager.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package token
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+func NewManager(c corev1.CoreV1Interface) *Manager {
+	return &Manager{
+		getToken: func(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+			return c.ServiceAccounts(namespace).CreateToken(name, tr)
+		},
+		cache: make(map[string]*authenticationv1.TokenRequest),
+		now:   time.Now,
+	}
+}
+
+type Manager struct {
+	getToken func(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error)
+
+	// cmu guards the cache
+	cmu   sync.RWMutex
+	cache map[string]*authenticationv1.TokenRequest
+
+	// mocked for testing
+	now func() time.Time
+}
+
+func (m *Manager) GetServiceAccountToken(name, namespace string, tr *authenticationv1.TokenRequest) (*authenticationv1.TokenRequest, error) {
+	key := keyFunc(name, namespace, tr)
+	ctr, ok := m.get(key)
+
+	var err error
+	if !ok || m.requiresRefresh(ctr) {
+		tr, err := m.getToken(name, namespace, tr)
+		if err != nil {
+			glog.Errorf("couldn't update token %s: %v", key, err)
+		} else {
+			m.set(key, tr)
+		}
+	}
+
+	ctr, ok = m.get(key)
+	if !ok && err != nil {
+		return nil, fmt.Errorf("failed to fetch initial token: %v", key, err)
+	}
+	if !ok {
+		// wat
+		return nil, fmt.Errorf("failed to fetch token")
+	}
+	if m.expired(ctr) && err != nil {
```

Comment:
There's too many cases here.
How about:
- get existing ctr
- if valid, return
- refresh ctr
- if error return
- store ctr
- return ctr

Reply:
changed code to be pretty close to this, except if there is an error refreshing the token but the old token hasn't expired, we should log an error and return the old token.